File: nand2tetris/projects/06/Rust/assembler/src/encoder.rs
use crate::Command;
use crate::Addr;
use crate::Dest;
use crate::{Comp, IR, R, AM};
use crate::Jump;

impl Command {
    pub fn encode(&self) -> u16 {
        match self {
            Command::A(Addr(v)) => *v,
            Command::C { dest, comp, jump } => {
                let c = match *comp {
                    Comp::O                         => 0b0101010, // "0"  
                    Comp::Id(IR::I)                 => 0b0111111, // "1"  
                    Comp::Neg(IR::I)                => 0b0111010, // "-1" 
                    Comp::Id(IR::R(R::D))           => 0b0001100, // "D"  
                    Comp::Id(IR::R(R::AM(AM::A)))   => 0b0110000, // "A"  
                    Comp::Not(R::D)                 => 0b0001101, // "!D" 
                    Comp::Not(R::AM(AM::A))         => 0b0110001, // "!A" 
                    Comp::Neg(IR::R(R::D))          => 0b0001111, // "-D" 
                    Comp::Neg(IR::R(R::AM(AM::A)))  => 0b0110011, // "-A" 
                    Comp::Inc(R::D)                 => 0b0011111, // "D+1"
                    Comp::Inc(R::AM(AM::A))         => 0b0110111, // "A+1"
                    Comp::Dec(R::D)                 => 0b0001110, // "D-1"
                    Comp::Dec(R::AM(AM::A))         => 0b0110010, // "A-1"
                    Comp::AddD(AM::A)               => 0b0000010, // "D+A"
                    Comp::SubD(AM::A)               => 0b0010011, // "D-A"
                    Comp::SubXD(AM::A)              => 0b0000111, // "A-D"
                    Comp::AndD(AM::A)               => 0b0000000, // "D&A"
                    Comp::OrD(AM::A)                => 0b0010101, // "D|A"
                    Comp::Id(IR::R(R::AM(AM::M)))   => 0b1110000, // "M"  
                    Comp::Not(R::AM(AM::M))         => 0b1110001, // "!M" 
                    Comp::Neg(IR::R(R::AM(AM::M)))  => 0b1110011, // "-M" 
                    Comp::Inc(R::AM(AM::M))         => 0b1110111, // "M+1"
                    Comp::Dec(R::AM(AM::M))         => 0b1110010, // "M-1"
                    Comp::AddD(AM::M)               => 0b1000010, // "D+M"
                    Comp::SubD(AM::M)               => 0b1010011, // "D-M"
                    Comp::SubXD(AM::M)              => 0b1000111, // "M-D"
                    Comp::AndD(AM::M)               => 0b1000000, // "D&M"
                    Comp::OrD(AM::M)                => 0b1010101, // "D|M"
                };
                let d = match *dest {
                    None => 0b000,
                    Some(Dest { a, d, m }) => {
                        [a, d, m].iter().fold(0b0, |acc, x| (acc << 1) + *x as u16)
                    }
                };
                let j = match *jump {
                    None => 0b000,
                    Some(Jump { lt, eq, gt }) => {
                        [lt, eq, gt].iter().fold(0b0, |acc, x| (acc << 1) + *x as u16)
                    }
                };
                (0b111 << 13) + [c, d, j].iter().fold(0b0, |acc, x| (acc << 3) + x)
            },
            _ => unreachable!()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bit_shift() {
        let c = 0b1010101;
        let d = 0b010;
        let j = 0b101;
        let trial = [c, d, j].iter().fold(0b0, |acc, x| (acc << 3) + x);
        let expect = 0b1010101_010_101;
        assert_eq!(trial, expect);
    }
    #[test]
    fn cmd_a() {
        let trial = Command::A(Addr(32767)).encode();
        let expect = 0b0111111111111111;
        assert_eq!(trial, expect);
    }
    #[test]
    fn cmd_c() {
        let c = Comp::OrD(AM::M);
        let d = Dest { a:false, d:true, m:false };
        let j = Jump { lt:true, eq:false, gt:true };
        let trial = Command::C { dest: Some(d), comp: c, jump: Some(j) }.encode();
        let expect = 0b111_1010101_010_101;
        assert_eq!(trial, expect);
    }
    #[test]
    fn bool_bit() {
        let trial = true as u16 ;
        let expect = 1;
        assert_eq!(trial, expect);
    }
}

File: nand2tetris/projects/06/Haskell/assembler/src/Encoder.hs
{-# LANGUAGE BinaryLiterals #-}

module Encoder ( encode ) where

import Numeric ( showIntAtBase )
import Data.Char ( intToDigit )
import Data.Bits ( Bits(shiftL) )
import qualified Data.Text as T

import Lib.Command as Cmd ( Command(C, A) )
import Lib.Addr as Addr ( Addr(..) )
import Lib.Dest as D ( Dest(..) )
import Lib.Comp as C ( Comp(..), IR(..), R(..), AM(..) )
import Lib.Jump as J ( Jump(..) )

type Binary = T.Text

encode :: Command -> Binary
encode c = T.pack . fill0 16 $ showIntAtBase 2 intToDigit (encode' c) ""
    where
        fill0 width = reverse . take width . (\s -> s ++ repeat '0') . reverse

encode' :: Command -> Int
encode' (Cmd.A (Addr a)) = a
encode' (Cmd.C d c j) = (0b111 `shiftL` 13) + foldl (\acc x -> (acc `shiftL` 3) + x) 0 [c', d', j']
    where
        c' = case c of
            C.O                     -> 0b0101010 -- "0"  
            C.Id C.I                -> 0b0111111 -- "1"  
            C.Neg C.I               -> 0b0111010 -- "-1" 
            C.Id (C.R C.D)          -> 0b0001100 -- "D"  
            C.Id (C.R (C.AM C.A))   -> 0b0110000 -- "A"  
            C.Not C.D               -> 0b0001101 -- "!D" 
            C.Not (C.AM C.A)        -> 0b0110001 -- "!A" 
            C.Neg (C.R C.D)         -> 0b0001111 -- "-D" 
            C.Neg (C.R (C.AM C.A))  -> 0b0110011 -- "-A" 
            C.Inc C.D               -> 0b0011111 -- "D+1"
            C.Inc (C.AM C.A)        -> 0b0110111 -- "A+1"
            C.Dec C.D               -> 0b0001110 -- "D-1"
            C.Dec (C.AM C.A)        -> 0b0110010 -- "A-1"
            C.AddD C.A              -> 0b0000010 -- "D+A"
            C.SubD C.A              -> 0b0010011 -- "D-A"
            C.SubXD C.A             -> 0b0000111 -- "A-D"
            C.AndD C.A              -> 0b0000000 -- "D&A"
            C.OrD C.A               -> 0b0010101 -- "D|A"
            C.Id (C.R (C.AM C.M))   -> 0b1110000 -- "M"  
            C.Not (C.AM C.M)        -> 0b1110001 -- "!M" 
            C.Neg (C.R (C.AM C.M))  -> 0b1110011 -- "-M" 
            C.Inc (C.AM C.M)        -> 0b1110111 -- "M+1"
            C.Dec (C.AM C.M)        -> 0b1110010 -- "M-1"
            C.AddD C.M              -> 0b1000010 -- "D+M"
            C.SubD C.M              -> 0b1010011 -- "D-M"
            C.SubXD C.M             -> 0b1000111 -- "M-D"
            C.AndD C.M              -> 0b1000000 -- "D&M"
            C.OrD C.M               -> 0b1010101 -- "D|M"
        d' = case d of
            Nothing   -> 0b000
            Just d -> case d of
                D.M   -> 0b001
                D.D   -> 0b010
                D.MD  -> 0b011
                D.A   -> 0b100
                D.AM  -> 0b101
                D.AD  -> 0b110
                D.AMD -> 0b111
        j' = case j of
            Nothing   -> 0b000
            Just j -> case j of
                J.JGT -> 0b001
                J.JEQ -> 0b010
                J.JGE -> 0b011
                J.JLT -> 0b100
                J.JNE -> 0b101
                J.JLE -> 0b110
                J.JMP -> 0b111
encode' _ = error "invalid command encode"
